File: tests/abi/midi_controllers_layout.cpp
#include "pluginterfaces/vst/ivstmidicontrollers.h"

#include <cstdio>

int main ()
{
	std::printf ("kCtrlBankSelectMSB %d\n", Steinberg::Vst::kCtrlBankSelectMSB);
	std::printf ("kCtrlModWheel %d\n", Steinberg::Vst::kCtrlModWheel);
	std::printf ("kCtrlBreath %d\n", Steinberg::Vst::kCtrlBreath);
	std::printf ("kCtrlFoot %d\n", Steinberg::Vst::kCtrlFoot);
	std::printf ("kCtrlPortaTime %d\n", Steinberg::Vst::kCtrlPortaTime);
	std::printf ("kCtrlDataEntryMSB %d\n", Steinberg::Vst::kCtrlDataEntryMSB);
	std::printf ("kCtrlVolume %d\n", Steinberg::Vst::kCtrlVolume);
	std::printf ("kCtrlBalance %d\n", Steinberg::Vst::kCtrlBalance);
	std::printf ("kCtrlPan %d\n", Steinberg::Vst::kCtrlPan);
	std::printf ("kCtrlExpression %d\n", Steinberg::Vst::kCtrlExpression);
	std::printf ("kCtrlBankSelectLSB %d\n", Steinberg::Vst::kCtrlBankSelectLSB);
	std::printf ("kCtrlSustainOnOff %d\n", Steinberg::Vst::kCtrlSustainOnOff);
	std::printf ("kCtrlSoundVariation %d\n", Steinberg::Vst::kCtrlSoundVariation);
	std::printf ("kCtrlEff1Depth %d\n", Steinberg::Vst::kCtrlEff1Depth);
	std::printf ("kCtrlAllSoundsOff %d\n", Steinberg::Vst::kCtrlAllSoundsOff);
	std::printf ("kAfterTouch %d\n", Steinberg::Vst::kAfterTouch);
	std::printf ("kPitchBend %d\n", Steinberg::Vst::kPitchBend);
	std::printf ("kCountCtrlNumber %d\n", Steinberg::Vst::kCountCtrlNumber);
	std::printf ("kCtrlProgramChange %d\n", Steinberg::Vst::kCtrlProgramChange);
	std::printf ("kSystemActiveSensing %d\n", Steinberg::Vst::kSystemActiveSensing);
	return 0;
}

File: tools/midi_controllers_layout.zig
const std = @import("std");
const midi = @import("vst3-zig").pluginterfaces.vst.ivstmidicontrollers;

pub fn main() !void {
    const stdout = std.io.getStdOut().writer();
    try printController(stdout, "kCtrlBankSelectMSB", midi.kCtrlBankSelectMSB);
    try printController(stdout, "kCtrlModWheel", midi.kCtrlModWheel);
    try printController(stdout, "kCtrlBreath", midi.kCtrlBreath);
    try printController(stdout, "kCtrlFoot", midi.kCtrlFoot);
    try printController(stdout, "kCtrlPortaTime", midi.kCtrlPortaTime);
    try printController(stdout, "kCtrlDataEntryMSB", midi.kCtrlDataEntryMSB);
    try printController(stdout, "kCtrlVolume", midi.kCtrlVolume);
    try printController(stdout, "kCtrlBalance", midi.kCtrlBalance);
    try printController(stdout, "kCtrlPan", midi.kCtrlPan);
    try printController(stdout, "kCtrlExpression", midi.kCtrlExpression);
    try printController(stdout, "kCtrlBankSelectLSB", midi.kCtrlBankSelectLSB);
    try printController(stdout, "kCtrlSustainOnOff", midi.kCtrlSustainOnOff);
    try printController(stdout, "kCtrlSoundVariation", midi.kCtrlSoundVariation);
    try printController(stdout, "kCtrlEff1Depth", midi.kCtrlEff1Depth);
    try printController(stdout, "kCtrlAllSoundsOff", midi.kCtrlAllSoundsOff);
    try printController(stdout, "kAfterTouch", midi.kAfterTouch);
    try printController(stdout, "kPitchBend", midi.kPitchBend);
    try printController(stdout, "kCountCtrlNumber", midi.kCountCtrlNumber);
    try printController(stdout, "kCtrlProgramChange", midi.kCtrlProgramChange);
    try printController(stdout, "kSystemActiveSensing", midi.kSystemActiveSensing);
}

fn printController(writer: anytype, comptime name: []const u8, controller: @TypeOf(midi.kCtrlBankSelectMSB)) !void {
    try writer.print("{s} {}\n", .{ name, controller });
}
